Tests and Bugfixes for the GB potential

Kischy opened the pull request:
Fixes #2922

Description of changes:
 - Added tests for the gb potential for the force and torque
 - Fixed bugs in gay_berne.hpp
 - Removed condition in gay_berne.hpp

PR Checklist
------------
 - [x] Tests?
   - [ ] Interface
   - [x] Core 
 - [ ] Docs?

Kischy committed:
Issue #2922: Refactoring energy tests gb potential

Kischy committed:
Issue #2922: Added tests for gb forces and corrected gay_berne.hpp. Removed condition.

Kischy committed:
Issue #2922: Added gb torque tests. Some refactoring

Kischy committed:
Issue #2922: Some refactoring

Kischy committed:
Issue #2922: Updated style to meet code formatting rules

Kischy committed:
#2922: Changed misleading variable name

jngrad committed:
Formatting

jngrad commented:
So far we've always used analytical solutions to calculate reference forces. Using a numerical approximation of the gradient is an interesting approach. I'm actually surprised that forces calculated with an integral step of 2e-7 can pass a `self.assertAlmostEqual` test at precision 1e-7. I've started the tests on non-amd64 architectures (pipeline [8879](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/pipelines/8879)) to see if there is a risk of numerical instability.

After re-organizing my version of the force function to match the nomenclature of the variables in the energy function, I still can't follow the force code. For example, in the calculation of epsilon, at line:
https://github.com/espressomd/espresso/blob/d79e9c14545827b07096a0e7af41227df48cb5e6/src/core/nonbonded_interactions/gay_berne.hpp#L70-L72
I would expect variable `E2` to be raised to the power of `mu - 1` in a subsequent variable to evaluate the derivative (d epsilon / d r), but I can't find it. It would be very helpful if we had the formula of the GB force, possibly with the intermediate steps to obtain it from the derivative of energy. We could write them with LaTeX in a Doxygen page, like we did for the [derivation of the bond angle force](http://espressomd.org/html/dox/bondedIA.html#bondedIA_angle_force).

Kischy commented:
> I've started the tests on non-amd64 architectures (pipeline 8879) to see if there is a risk of numerical instability.

I saw that two of them failed. This [one](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/152120) failed because the reaction_ensemble test had a timeout. The [second one](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/152124) failed because of different timeouts and fails, non of which was the interactions_non-bonded.py. Do you know why those failed?

> So far we've always used analytical solutions to calculate reference forces. Using a numerical approximation of the gradient is an interesting approach. I'm actually surprised that forces calculated with an integral step of 2e-7 can pass a self.assertAlmostEqual test at precision 1e-7.

I think this way testing derivatives can be very powerful, if the analytical solution is long or complicated. Although it is much slower to calculate the derivative this way, it is not so prone to error. To advertise the approach a little bit more: If I had just implemented the same analytical solution used in the c++ code for testing I would not have found the error. I borrowed the idea from machine learning, see [here](http://ufldl.stanford.edu/tutorial/supervised/DebuggingGradientChecking/) and [here](https://stackoverflow.com/questions/47506521/what-exactly-is-gradient-checking).

> I would expect variable E2 to be raised to the power of mu - 1 in a subsequent variable to evaluate the derivative (d epsilon / d r), but I can't find it. 

It is a little bit confusing. The change in power comes with this line: https://github.com/espressomd/espresso/blob/d79e9c14545827b07096a0e7af41227df48cb5e6/src/core/nonbonded_interactions/gay_berne.hpp#L77

And with this line:
https://github.com/espressomd/espresso/blob/d79e9c14545827b07096a0e7af41227df48cb5e6/src/core/nonbonded_interactions/gay_berne.hpp#L78

And also with these lines:
https://github.com/espressomd/espresso/blob/d79e9c14545827b07096a0e7af41227df48cb5e6/src/core/nonbonded_interactions/gay_berne.hpp#L89-L90

> It would be very helpful if we had the formula of the GB force, possibly with the intermediate steps to obtain it from the derivative of energy.

If you want I can send you my jupyter notebook (using python 3) or the python script, where I did the calculations of the derivative. 

>  We could write them with LaTeX in a Doxygen page, like we did for the derivation of the bond angle force.

The problem is that I do not know what paper the orginal author used, so I can't cite anything. I will try to work out a new nomenclatur for the problem and then we can put it in latex code, if you want.

jngrad committed:
Refactor Gay-Berne force/energy kernels

* add extra intermediate variables
* use same variable names for identical or very similar
  intermediate quantities in the force/energy kernels
* simplify code logic with vector/scalar utility functions

jngrad commented:
The non-amd tests are known to sometimes false trigger, that's why we don't run them automatically. The gradient approach seems to work, though. Your [UFLDL link](http://ufldl.stanford.edu/tutorial/supervised/DebuggingGradientChecking/) states that an epsilon of 1e-4 will usually provide results at 4 digits precision, but in my experience a test is safer when the reference force is at least one order of magnitude more precise than the test threshold delta. Maybe @KaiSzuttor can comment on that.

The book or article used by the original author is unknown, all we know is this code is 16 years old. Given the notation in the energy function and the [LaTeX formula in the docs](http://espressomd.org/html/doc/inter_non-bonded.html#gay-berne-interaction), I would bet on [Cleaver *et al.* 1996](https://doi.org/10.1103/PhysRevE.54.559), equations 3-4 and 7-9.

Here is [my modified version](https://github.com/jngrad/espresso/blob/Bugfix_Tests_Gay_Berne_Issue_2922-nomencl/src/core/nonbonded_interactions/gay_berne.hpp) of the GB force function. Although it looks more verbose, when compiled with `-O3` it generates 23 less lines of assembly code in gcc and 13 less in Clang (9% resp. 6% less). We could probably reduce the assembly code further with more refactoring based on the analytical expression of the force. I've looked at several sources on the GB potential, but none of them give the expression of its derivative in Cartesian space. If you already worked out the derivation, you can share it here by attaching a file.

Kischy commented:
> The non-amd tests are known to sometimes false trigger, that's why we don't run them automatically. The gradient approach seems to work, though. Your UFLDL link states that an epsilon of 1e-4 will usually provide results at 4 digits precision, but in my experience a test is safer when the reference force is at least one order of magnitude more precise than the test threshold delta. Maybe @KaiSzuttor can comment on that.

Okay, I will change the epsilon of my force calculation and update the pull request.

> Here is my modified version of the GB force function. Although it looks more verbose, when compiled with -O3 it generates 23 less lines of assembly code in gcc and 13 less in Clang (9% resp. 6% less). We could probably reduce the assembly code further with more refactoring based on the analytical expression of the force. I've looked at several sources on the GB potential, but none of them give the expression of its derivative in Cartesian space. If you already worked out the derivation, you can share it here by attaching a file.

I worked on a better nomenclature just now too. If it is okay with you I would combine my version with yours and we see how it fits. _What do you think?_

Kischy commented:
I can't attach the files here, because the file type is not supported by github. I will send them to you by mail.

Kischy committed:
#2922: Changed delta value

Kischy commented:
> Here is my modified version of the GB force function.

I think you're version is the best we can do. Can you just add the changed version to my pull request or should I do so?

Kischy commented:
> Your UFLDL link states that an epsilon of 1e-4 will usually provide results at 4 digits precision, but in my experience a test is safer when the reference force is at least one order of magnitude more precise than the test threshold delta. 

I added a commit that changes the delta value to 1x10E-6. For some reason it does not show here yet.

jngrad commented:
You can use my version. The simplest way to do it is to merge my branch into your branch (this is a trivial fast-forward), then add the new commit that changes the precision. Doing it the other way around (adding the precision commit, then merging my branch) would require a complex merge commit.

Also, I would not change delta to a bigger value, instead I would change epsilon to 1x10E-8.

jngrad commented:
You probably forgot to push the new commit to the GitHub repository. If it's ok with you, I can add my new version into your branch on GitHub. Then on you computer, you'll have to discard the precision commit (for example with `git reset --soft HEAD^` to rewind the branch history to d79e9c1), then pull your GitHub branch into your local repository.

Kischy commented:
No, it is already pushed. I'll merge your commit, change the delta and epsilon and push again.

Kischy committed:
Resolved small merge conflict

Kischy committed:
Merge fixes of gay berne potential

Kischy commented:
> Also, I would not change delta to a bigger value, instead I would change epsilon to 1x10E-8.

The value of 1x10E-8 does not work, probably because of the numerical errors mentioned in the [UFLDL link ](http://ufldl.stanford.edu/tutorial/supervised/DebuggingGradientChecking/). I think 1x10E-7 is the lowest we can go here. Therefore I would recommend changing the delta value to 1x10E-6. Is that fine with you or not?

KaiSzuttor commented on testsuite/python/tests_common.py:
```diff
@@ -596,6 +596,20 @@ def gay_berne_potential(r_ij, u_i, u_j, epsilon_0, sigma_0, mu, nu, k_1, k_2):
     return 4. * epsilon * (rr**-12 - rr**-6)
 
 
+def calc_derivative(func, x, eps=1.0e-7, index=None):
```
could you please use [```scipy.misc.derivative```](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.derivative.html). If that does not suit your needs you should definitely add a docstring to this function. Actually we could even use [```sympy```](https://www.sympy.org/en/index.html). But we probably don't want to depend on that.

Kischy replied:
I will add a docstring. [scipy.misc.derivative](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.derivative.html) does not allow for a multidimensional input.

KaiSzuttor replied:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.vectorize.html

Kischy replied:
Does that help? I need to manipulate the element at a specific index while the rest stays the same.

Kischy replied:
I added a docstring. I could have thrown an error if someone gives a multidimensional input and index=None, but I think it would be a "overkill". In the docstring it is stated, that the index needs to be specified if x is an array.

jngrad commented:
I've just discussed it with Kai, and we're both fine with a larger delta.

Kischy commented:
> 
> 
> I've just discussed it with Kai, and we're both fine with a larger delta.

Thanks for the update.

Kischy committed:
#2922:
Added docstring to calc_derivative
changed delta value in test_gb

Kischy commented:
Everything should be okay now. Do you need something else concerning issue #2922?

Kischy committed:
Merge branch 'python' into Bugfix_Tests_Gay_Berne_Issue_2922

jngrad commented:
Note: it's unnecessary to merge python into this branch in the absence of merge conflicts, we have a bot to do that during the merge.

Kischy commented:
> 
> 
> Note: it's unnecessary to merge python into this branch in the absence of merge conflicts, we have a bot to do that during the merge.

Ok, sorry about that.

KaiSzuttor commented:
I think the confusion arises from the fact that the ```tests_common.gay_berne_potential``` returns an array and not a scalar as one would expect from a potential. I think we should not try to create a special derivative function that handles this case.

jngrad commented:
I'm also confused by the derivative function, in that it's actually a partial derivative function that expects a single N-dimensional coordinate, while scipy.misc.derivative expects a N-dimensional array of points scattered on the x-axis. The need for a partial derivative function comes from the multi-dimensional nature of the input values for the GB energy kernel. I'm currently rewriting the derivative function. I couldn't find an equivalent function in numpy and scipy.

KaiSzuttor commented:
can you explain why the potential is not a scalar?

Kischy commented:
The potential is scalar.

jngrad commented:
The potential is a scalar (linalg.norm without axis returns a scalar):
https://github.com/espressomd/espresso/blob/a910bab0e1c64e4fb443e76af546e4570253bb39/testsuite/python/tests_common.py#L594-L596
The input however isn't.

Kischy commented:
> I'm also confused by the derivative function, in that it's actually a partial derivative function that expects a single N-dimensional coordinate, while scipy.misc.derivative expects a N-dimensional array of points scattered on the x-axis. The need for a partial derivative function comes from the multi-dimensional nature of the input values for the GB energy kernel. I'm currently rewriting the derivative function. I couldn't find an equivalent function in numpy and scipy.

You are right. It is a partial derivative function. Did you already start the rewriting? I did the same thing half an hour ago.

KaiSzuttor commented:
this is not a scalar

https://github.com/espressomd/espresso/blob/a910bab0e1c64e4fb443e76af546e4570253bb39/testsuite/python/tests_common.py#L588

right?

Kischy commented:
> 
> 
> this is not a scalar
> 
> https://github.com/espressomd/espresso/blob/a910bab0e1c64e4fb443e76af546e4570253bb39/testsuite/python/tests_common.py#L588
> 
> right?

No, this gives a scalar for epsilon.

KaiSzuttor commented:
maybe we should save developer time and talk before code :smile:

Kischy commented:
Yeah, thats a good idea. :D

jngrad commented:
I'll upload my version of the partial derivative in a separate branch.

Kischy commented:
Okay, than I will upload mine here and you can pick which one you like better.

jngrad commented:
Here is my version: [calc_partial_derivative()](https://github.com/jngrad/espresso/blob/9cac5045d450d27d6f6127245ce8443f19c7138e/testsuite/python/tests_common.py#L599)

Kischy commented:
> 
> 
> Here is my version: [calc_partial_derivative()](https://github.com/jngrad/espresso/blob/9cac5045d450d27d6f6127245ce8443f19c7138e/testsuite/python/tests_common.py#L599)

Okay, I think a combination of our versions might be best. Mine calculates all the derivatives in one go. _What do you think?_ (Before I start coding again :D)
Nice Docstring by the way.

jngrad commented:
What do you mean? The unit test hasn't changed: [9cac504::interactions_non-bonded.py](https://github.com/jngrad/espresso/commit/9cac5045d450d27d6f6127245ce8443f19c7138e#diff-8080230a7bbebfde7eab64f61e78f499). The new version can also compute the derivative over two dimensions at once, e.g. `f(x,y)=(x^2+y^2)/2`, gradient: `df(x,y)=x+y`, implementation: `calc_partial_derivative(lambda x: 0.5*x[0]**2 + 0.5*x[1]**2, [1,1], variables=[0,1])` returns 2.0

Kischy commented:
I changed the test functions and put a for loop into the function to calculate all the partial derivatives of the function at point x0 at once.

Kischy commented:
It's just another implementation, I am fine with yours if you want. Can you merge it into this pull request or do I need to do that?

jngrad commented:
We should first decide on the interface. Currently my function works for a single coordinate in N-dimensional space, whereas yours seems to work on a numpy array of points in N-dimensional space. That would be an extension of scipy.misc.derivative. Is that correct?

Kischy commented:
I would go for a implementation like this:

```python
def calc_partial_derivatives(func, x0, dx=1.0e-7):    
    """
    Approximate all the partial derivatives of a function at a point x0 
    using the two-point central difference formula with spacing dx.

    Parameters
    ----------
    func: `function` obj
        function from which the derivative is calculated
    x0: (N) array_like of obj:`float`
        Point in N-dimensional space where the derivatives are calculated
    dx: :obj:`float`, optional
        Spacing
    Returns
    -------
    array_like of obj:`float` or `float` (if x0 is unidimensional)
        the partial derivatives    
    """       
    
    unidimensional = not np.array(x0).shape
    if unidimensional:
        x0 = [x0]        
        
    derivs = []
    
    for i, _ in enumerate(x0):
        x_plus = np.array(x0, dtype=float)
        x_minus = np.array(x0, dtype=float)        
        x_plus[i] += dx
        x_minus[i] -= dx
        
        derivs.append((func(x_plus) - func(x_minus)) / (2.0 * dx))    
        
    if unidimensional:
        derivs = derivs[0]
    else:
        derivs = np.array(derivs)

    return derivs
```


This is a combination of both our implementations.

Kischy commented:
> That would be an extension of scipy.misc.derivative. 

I'm not sure. I would say an extension of the scipy.misc.derivative would be a vectorized version. As you said before this function calculates partial derivatives and therefore it does something different, I think.

Kischy commented:
```python
def calc_partial_derivatives(func, x0, dx=1.0e-7):    
    """
    Approximate all the partial derivatives of a function at a point x0 
    using the two-point central difference formula with spacing 2dx.

    Parameters
    ----------
    func: :obj:`function`
        function from which the derivative is calculated
    x0: (N,) array_like of :obj:`float`
        Point in N-dimensional space where the derivatives are calculated
    dx: :obj:`float`, optional
        Spacing
    Returns
    -------
    array_like of obj:`float` or :obj:`float` (if x0 is unidimensional)
        the partial derivatives    
    """       
    
    unidimensional = not np.array(x0).shape
    if unidimensional:
        x0 = [x0]        
        
    derivs = []
    
    for i, _ in enumerate(x0):
        x_plus = np.array(x0, dtype=float)
        x_minus = np.array(x0, dtype=float)        
        x_plus[i] += dx
        x_minus[i] -= dx
        
        derivs.append((func(x_plus) - func(x_minus)) / (2.0 * dx))    
        
    if unidimensional:
        derivs = derivs[0]
    else:
        derivs = np.array(derivs)

    return derivs
```

Kischy committed:
#2922:
Changed calc_derivative to calc_partial_derivatives and moved it to
the test_gb function. Some refactoring

Kischy commented:
Styling rules again. Sorry.

Kischy committed:
#2922: Applied styling rules

KaiSzuttor commented on testsuite/python/interactions_non-bonded.py:
```diff
@@ -588,40 +588,167 @@ def test_gaussian(self):
     def test_gb(self):
+
+        # helper function definitions
+        def calc_partial_derivatives(func, x0, dx=1.0e-7):    
```
aka gradient

Kischy replied:
Ok, did you rename it and update the docstring?

jngrad replied:
Yes: 377b216

Kischy replied:
ty

jngrad committed:
Rename gradient function

jngrad commented:
bors r+

KaiSzuttor committed:
testsuite: improve readability of gradient calculation.

KaiSzuttor committed:
Merge branch 'Bugfix_Tests_Gay_Berne_Issue_2922' of github.com:Kischy/espresso into Bugfix_Tests_Gay_Berne_Issue_2922

KaiSzuttor committed:
numpy as np.

KaiSzuttor commented:
I was not yet done...

KaiSzuttor commented:
I think we should have all reviewers approve the PR before merging

Kischy commented:
You changed the interface of the gradient function so that `x0: (3,)`? Does that matter that much now that it is specific to the gb_test?

Could you also update the Docstring?
`Point in N-dimensional space where the derivatives are calculated`
to
`Point in 3-dimensional space where the derivatives are calculated`

KaiSzuttor commented:
the scope of the function is limited to ```test_gb```. Have you ever calculated a gradient in more than three dimensions?

Kischy commented:
> the scope of the function is limited to test_gb.

That's why I thought that it doesn't matter.

> Have you ever calculated a gradient in more than three dimensions?

No

KaiSzuttor commented:
I thought it's just more readable

Kischy commented:
Yeah, you are right.

Kischy commented:
Can you exchange every "derivative" with "gradient" in the Docstring? I think than we are done. :)

KaiSzuttor committed:
testsuite: docstring.

KaiSzuttor commented:
sorry for the long discussion, but I think we overall improved readability. Thanks for your contribution!

Kischy commented:
@KaiSzuttor  I like your implementation of the gradiant function, it is quite compact. However it might be slower, which does not matter in this case. Just keep that in mind, if you ever use it heavily somewhere else.

Kischy commented:
> 
> 
> sorry for the long discussion, but I think we overall improved readability. Thanks for your contribution!

Yes, it looks better now. Thank you. I'm glad I could help. And no worries about the discussion.

KaiSzuttor commented:
> However it might be slower

Why?

Kischy commented:
> > However it might be slower
> 
> Why?

There will be one lambda for each call to the function. That will make it slower in comparison. I shouldn't have brought it up, because it does not matter.

KaiSzuttor commented:
bors r+
